core/src/main/java/org/geysermc/geyser/entity/type/player/PlayerEntity.java

```diff
@@ -422,6 +423,32 @@ protected void scoreVisibility(boolean show) {
         dirtyMetadata.put(EntityDataTypes.SCORE, show ? cachedScore : "");
     }
 
+    @Override
+    public void setPose(Pose pose) {
+        super.setPose(pose);
+        setFlag(EntityFlag.SWIMMING, false);
+        setFlag(EntityFlag.CRAWLING, false);
+
+        if (pose == Pose.SWIMMING) {
+            // This is just for, so we know if player is swimming or crawling.
+            if (session.getGeyser().getWorldManager().blockAt(session, this.position().toInt()).is(Blocks.WATER)) {
+                setFlag(EntityFlag.SWIMMING, true);
+            } else {
+                setFlag(EntityFlag.CRAWLING, true);
+            }
+        }
+
+        // Look at https://github.com/GeyserMC/Geyser/issues/5316, we're fixing this by spoofing player pitch to 0.
+        if (getFlag(EntityFlag.CRAWLING)) {
+            updateRotation(this.yaw, 0, this.onGround);
+        }
```

Comment:
This could probably be moved up into the if-statement above, as the crawling flag is false anyways since that was reset at the top